### PR TITLE
Fixed documentation of GaussProductExperiment

### PR DESCRIPTION
**Pull request by mbaudin47**
The goal of this PR is to fix the documentation of `GaussProductExperiment`. Indeed, the number of nodes for each marginal has name "MarginalDegree", which is a misnomer. Indeed, this is not the degree of a polynomial. Moreover, it is not the marginal degree of exactness. In order to keep in consistent with the name used in other experiment, I suggest to use the name "size". 
- Replaced `MarginalDegrees` with `MarginalSizes`. 
- Added theoretical doc. 
- Added example.

One of the goals of this PR is to see how this class is consistent with the new `TensorProductExperiment` created at #1963. It seems to me that there is a great amount of duplication, both at the C++ source code level, unit test and doc. In order to make the design simpler and consistent, I suggest to create a new `UnivariateGaussExperiment`, which takes a distribution with one dimension and returns the corresponding univariate Gauss experiment. This is to be discussed before moving on to coding step.

### CODE REVIEW

**Comment by mbaudin47** [comment]
This is not finished yet; the `GaussProductExperiment` is also used in some examples:
- [Examples : Advanced polynomial chaos construction](https://8637-40717169-gh.circle-artifacts.com/0/html/auto_meta_modeling/polynomial_chaos_metamodel/plot_functional_chaos_advanced_ctors.html?highlight=gaussproductexperiment)
- [Examples : Create a Gauss product design](https://8637-40717169-gh.circle-artifacts.com/0/html/auto_reliability_sensitivity/design_of_experiments/plot_gauss_product_experiment.html?highlight=gaussproductexperiment)
- [Examples : Create a polynomial chaos metamodel by integration on the cantilever beam](https://8637-40717169-gh.circle-artifacts.com/0/html/auto_meta_modeling/polynomial_chaos_metamodel/plot_chaos_cantilever_beam_integration.html?highlight=gaussproductexperiment)

File: python/doc/bibliography.rst [hunk 1]
```diff
@@ -74,6 +74,8 @@ Bibliography
 .. [gamboa2013] Gamboa, F., Janon, A., Klein, T. & Lagnoux, A. *Sensitivity
     analysis for multidimensional and functional outputs.* 2013.
     `pdf <https://arxiv.org/pdf/1311.1797.pdf>`__
+.. [gerstner1998] Gerstner, T., & Griebel, M. (1998). *Numerical integration using
+    sparse grids.* Numerical algorithms, 18 (3), 209-232.
```
Comment (josephmure) [inline python/doc/bibliography.rst:78]:
```suggestion
    sparse grids.* Numerical algorithms, 18 (3), 209-232.
    `pdf <https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.33.3141&rep=rep1&type=pdf>`__
```

File: python/doc/bibliography.rst [hunk 2]
```diff
@@ -167,6 +169,8 @@ Bibliography
     computational experiments.*
     Journal of Statistical Planning and Inference, 43 :381-402, 1995.
     `pdf <https://www.osti.gov/servlets/purl/10184343>`__
+.. [morokoff1995] Morokoff, W. J., & Caflisch, R. E. (1995). Quasi-monte carlo 
+    integration. Journal of computational physics, 122(2), 218-230.
```
Comment (josephmure) [inline python/doc/bibliography.rst:173]:
```suggestion
.. [morokoff1995] Morokoff, W. J., & Caflisch, R. E. (1995). *Quasi-Monte Carlo 
    integration.* Journal of computational physics, 122(2), 218-230.
```

File: python/doc/examples/meta_modeling/polynomial_chaos_metamodel/plot_chaos_cantilever_beam_integration.py [hunk 3]
```diff
@@ -73,12 +73,12 @@
 # We begin by getting the standard measure associated with the multivariate polynomial basis. We see that the range of the `Beta` distribution has been standardized into the [-1,1] interval. This is the same for the `Uniform` distribution and the second `Beta` distribution.
 
 # %%
-distributionMu = multivariateBasis.getMeasure()
-distributionMu
+distributionStandard = multivariateBasis.getMeasure()
+distributionStandard
```
Comment (josephmure) [inline python/doc/examples/meta_modeling/polynomial_chaos_metamodel/plot_chaos_cantilever_beam_integration.py:77]:
```suggestion
print(distributionStandard)
```
Otherwise the output is presented as normal text.

File: python/doc/examples/meta_modeling/polynomial_chaos_metamodel/plot_functional_chaos_advanced_ctors.py [hunk 4]
```diff
@@ -143,8 +143,8 @@
 # Get the measure mu associated to the multivariate basis.
 
 # %%
-distributionMu = multivariateBasis.getMeasure()
-distributionMu
+distributionStandard = multivariateBasis.getMeasure()
+distributionStandard
```
Comment (josephmure) [inline python/doc/examples/meta_modeling/polynomial_chaos_metamodel/plot_functional_chaos_advanced_ctors.py:147]:
```suggestion
print(distributionStandard)
```
Same.

File: python/src/GaussProductExperiment_doc.i.in [hunk 5]
```diff
@@ -2,25 +2,87 @@
...
+    `GaussProductExperiment-DefaultMarginalSize` key of the :class:`~openturns.ResourceMap`. 
 distribution :
-    Distribution :math:`\mu` of dimension :math:`n` with an independent copula.
+    Distribution :math:`\mu` of dimension :math:`n_x` with an independent copula.
 
 Notes
 -----
-The number of points for each direction doesn't change and the :meth:`generate`
-method always gives the same sample.
+The tensor product design of experiments (DOE) is based on a collection of marginal multidimensional elementary design of experiments. It is anisotropic in the sense that each marginal DOE does not necessarily have the same size. Using more points in one component may be useful if we want to get more accurate results in that particular direction.
```
Comment (josephmure) [inline python/src/GaussProductExperiment_doc.i.in:24]:
```suggestion
The tensor product design of experiments (DOE) is based on a collection of marginal multidimensional elementary designs of experiments. It is anisotropic in the sense that each marginal DOE does not necessarily have the same size. Using more points in one component may be useful if we want to get more accurate results in that particular direction. 
```

File: python/src/GaussProductExperiment_doc.i.in [hunk 6]
```diff
@@ -2,25 +2,87 @@
...
+    Q_{s_1}^{(1)} \otimes \cdots \otimes Q_{s_{n_x}}^{(1)} (g) = \sum_{k_1 = 1}^{s_1} \cdots \sum_{k_{n_x} = 1}^{s_{n_x}} w_{s_1, k_1} \cdots w_{s_{n_x}, k_{n_x}}  g\left(x_{s_1, k_1}, \ldots, x_{s_{n_x}, k_{n_x}}\right).
+
+For any multi-index :math:`\boldsymbol{k} = (k_1, ..., k_{n_x}) \in (\mathbb{N}^\star)^{n_x}`, we write 
+
+.. math::
+
+    \boldsymbol{k} \leq \left(s_1, ..., s_{n_x}\right)
+
+if and only if :math:`k_j \leq s_j` for :math:`j = 1, ..., n_x`. This means that each component of the multi-index is lower than its corresponding marginal size.
+The set of multi-indices such that :math:`\boldsymbol{k} \leq \left(s_1, ..., s_{n_x}\right)` is produced using all the possible combinations of the indices with the :class:`~openturns.Tuples` class.
```
Comment (josephmure) [inline python/src/GaussProductExperiment_doc.i.in:66]: I find this sentence unclear. By definition, this set contains all possible combinations of the indices that fit the constraint. What are you trying to say?

**Review by josephmure — COMMENTED** [review]
Good job! I especially liked the new API doc, which is much more informative. Only minor remarks.

File: python/doc/bibliography.rst (hunk 2 above)
Comment (jschueller) [reply python/doc/bibliography.rst:173]: https://www.math.ucla.edu/~caflisch/Pubs/Pubs1995-1999/actaNumerica1998.pdf

File: lib/src/Uncertainty/Algorithm/WeightedExperiments/GaussProductExperiment.cxx [hunk 7]
```diff
@@ -210,16 +210,16 @@ void GaussProductExperiment::save(Advocate & adv) const
...
 }
 
 /* Method load() reloads the object from the StorageManager */
 void GaussProductExperiment::load(Advocate & adv)
 {
   WeightedExperimentImplementation::load(adv);
   adv.loadAttribute("collection_", collection_);
-  adv.loadAttribute("marginalDegrees_", marginalDegrees_);
-  setDistributionAndMarginalDegrees(distribution_, marginalDegrees_);
+  adv.loadAttribute("marginalSizes_", marginalSizes_);
```
Comment (jschueller) [inline lib/src/Uncertainty/Algorithm/WeightedExperiments/GaussProductExperiment.cxx:221]: you'll want to provide a backward-compatible load path here:
```
if (adv.hasAtribute("marginalSizes_"))
  adv.loadAttribute("marginalSizes_", marginalSizes_);
else
adv.loadAttribute("marginalDegrees_", marginalDegrees_);
```

**Review by jschueller — APPROVED** [review]
ok provided you add a backward-compatible load path

File: python/src/GaussProductExperiment_doc.i.in [hunk 8]
```diff
@@ -2,25 +2,87 @@
...
+
+.. math::
+
+    \int_{\mathcal{X}} g(\boldsymbol{x}) f(\boldsymbol{x}) d\boldsymbol{x} \approx \sum_{i = 1}^{s_t} w_i g\left(\boldsymbol{x}_i\right)
+
+where :math:`s_t \in \mathbb{N}` is the size of the tensor product design of experiments, :math:`w_1, ..., w_{s_t} \in \mathbb{R}` are the weights and :math:`\boldsymbol{x}_1, ..., \boldsymbol{x}_{s_t} \in \mathbb{R}^{n_x}` are the nodes. 
+
+Let :math:`\boldsymbol{k} = \left(k_1, ..., k_{n_x}\right) \in (\mathbb{N}^\star)^{n_x}`  be a multi-index where :math:`\mathbb{N}^\star = \{1, 2, ..., \}` is the set of natural numbers without zero. For each marginal index :math:`j = 1, ..., n_x`, let :math:`s_j \in \mathbb{N}` be its size, representing the number of nodes in the marginal quadrature. Let :math:`\left\{ Q_{s_j}^{(1)} \right\}_{j = 1, ..., n_x}` be the collection of marginal experiments where the exponent :math:`(1)` denotes univariate quadrature rule.
```
Comment (jschueller) [inline python/src/GaussProductExperiment_doc.i.in:39]: most of these lines are way too long and do not ease the review, please skip lines

File: python/doc/bibliography.rst (hunk 2 above)
Comment (mbaudin47) [reply python/doc/bibliography.rst]: Thank you for the link. I added it.

File: python/src/GaussProductExperiment_doc.i.in (hunk 6 above)
Comment (mbaudin47) [reply python/src/GaussProductExperiment_doc.i.in:66]: I want to write the word "combination" in plain english, and point to the `Tuples` class.

File: lib/src/Uncertainty/Algorithm/WeightedExperiments/GaussProductExperiment.cxx (hunk 7 above)
Comment (mbaudin47) [reply lib/src/Uncertainty/Algorithm/WeightedExperiments/GaussProductExperiment.cxx:221]: Fine. Fixed.

File: python/src/GaussProductExperiment_doc.i.in (hunk 6 above)
Comment (mbaudin47) [reply python/src/GaussProductExperiment_doc.i.in:66]: I consider this as "resolved", but feel free to reopen it if required.

File: python/src/GaussProductExperiment_doc.i.in (hunk 8 above)
Comment (mbaudin47) [reply python/src/GaussProductExperiment_doc.i.in]: Ok. Fixed.

**Comment by mbaudin47** [comment]
The draft seems to be finished. The decision has to be made.
- Should the `GaussProductExperiment` class disappear? Pros: it simplifies the code, the doc and the tests. All of them are at least in part duplicated. Cons: it breaks backward compatibility.
- Is the `UnivariateProductExperiment` the way to go? Pros: it fits in the `TensorProductExperiment` design. Pros : it would be compared with the `UnivariateFejerExperiment` which is on its way by @efekhari27. Cons: it forces to always use `TensorProductExperiment`, which creates an extra line of code. 

Creating a draft Python script may help to decide which design is best.

**Comment by mbaudin47** [comment]
Here is an example of both APIs.
```python
import openturns as ot

# Generate a tensorized Gauss-Legendre rule in 2 dimensions.
# The first marginal has 3 nodes.
# The first marginal has 5 nodes.

# 1. With TensorProductExperiment
experiment1 = ot.GaussProductExperiment(ot.Uniform(0.0, 1.0), [3])
experiment2 = ot.GaussProductExperiment(ot.Uniform(0.0, 1.0), [5])
collection = [experiment1, experiment2]
experiment = ot.TensorProductExperiment(collection)
sample, weights = experiment.generateWithWeights()
print("size = ", sample.getSize())

# 2. With GaussProductExperiment
distribution = ot.ComposedDistribution([ot.Uniform(0.0, 1.0)] * 2)
marginalSizes = [3, 5]
experiment = ot.GaussProductExperiment(distribution, marginalSizes)
sample, weights = experiment.generateWithWeights()
print("size = ", sample.getSize())
```

We see that the `TensorProductExperiment` actually uses `GaussProductExperiment`. This is why both classes are relevant and I set this PR as "ready for review".

File: python/src/GaussProductExperiment_doc.i.in [hunk 9]
```diff
@@ -2,25 +2,114 @@
...
 
 Parameters
 ----------
-marginalDegrees : sequence of positive int
-    Numbers of points :math:`N_i` for each direction. Then, the total number of
-    points generated is :math:`cardI=\prod_{i=1}^n N_i`. If not specified,
-    each value :math:`N_i` is :math:`5`.
+marginalSizes : sequence of positive int
+    Numbers of points :math:`s_j` for each direction. Then, the total number of
+    points generated is :math:`\textrm{card}(I)=\prod_{j = 1}^{n_x} s_j`.
```
Comment (josephmure) [inline python/src/GaussProductExperiment_doc.i.in:15]:
```suggestion
    points generated is :math:`\textrm{card}(I)=\prod_{j = 1}^{d_x} s_j`. 
```
To be consistent with the `TensorProductExperiment` doc.

File: python/src/GaussProductExperiment_doc.i.in [hunk 10]
```diff
@@ -2,25 +2,114 @@
...
-    each value :math:`N_i` is :math:`5`.
+marginalSizes : sequence of positive int
+    Numbers of points :math:`s_j` for each direction. Then, the total number of
+    points generated is :math:`\textrm{card}(I)=\prod_{j = 1}^{n_x} s_j`. 
+    By default, the value of :math:`s_j` is equal to :math:`5`.
+    The default marginal size is defined in the 
+    `GaussProductExperiment-DefaultMarginalSize` key of the :class:`~openturns.ResourceMap`. 
 distribution :
-    Distribution :math:`\mu` of dimension :math:`n` with an independent copula.
+    Distribution :math:`\mu` of dimension :math:`n_x` with an independent copula.
```
Comment (josephmure) [inline python/src/GaussProductExperiment_doc.i.in:20]:
```suggestion
    Distribution :math:`\mu` of dimension :math:`d_x` with an independent copula.
```
Same.

File: python/src/GaussProductExperiment_doc.i.in [hunk 11]
```diff
@@ -2,25 +2,114 @@
...
+.. math::
+
+    w_{\boldsymbol{k}} = \prod_{j = 1}^{n_x} w_{s_j, k_j} \in \mathbb{R}
+
+and each node is equal to the agregation of the marginal elementary nodes:
+
+.. math::
+
+    \boldsymbol{x}_{\boldsymbol{k}}
+    = \left(x_{s_1, k_1}, \ldots, x_{s_{n_x}, k_{n_x}}\right) \in \mathbb{R}^{n_x}.
```
Comment (josephmure) [inline python/src/GaussProductExperiment_doc.i.in:112]: This doc page could rely on the `TensorProductExperiment` doc from PR #1963.
```suggestion
The Gauss product experiment is a tensor product experiment
which uses Gauss points in each direction
(so, using the notations of the :class:`~openturns.TensorProductExperiment`
documentation, :math:`n_e = d_x` and :math:`d_j=1`
for every index :math:`j = 1, ..., n_e`).

For each marginal, the algorithm computes the family of orthogonal 
polynomials depending on the marginal distribution using the
:class:`~openturns.StandardDistributionPolynomialFactory` class. 
The input distribution must have an independent copula.  
```

File: python/src/GaussProductExperiment_doc.i.in [hunk 12]
```diff
@@ -2,25 +2,114 @@
...
 
 See also
 --------
```
Comment (josephmure) [inline python/src/GaussProductExperiment_doc.i.in:115]: This is out of reach of GitHub suggestions, but I would add `TensorProductExperiment` to the "See also" section.

**Comment by mbaudin47** [comment]
I wait for #1963 to be merged, so that this one can be updated and use TensorProductExperiment.

File: python/src/GaussProductExperiment_doc.i.in (hunk 12 above)
Comment (mbaudin47) [reply python/src/GaussProductExperiment_doc.i.in:36]: Fixed.

File: python/src/GaussProductExperiment_doc.i.in [hunk 13]
```diff
@@ -2,61 +2,107 @@
...
+
+For each marginal, the algorithm computes the family of orthogonal 
+polynomials depending on the marginal distribution using the
+:class:`~openturns.StandardDistributionPolynomialFactory` class. 
+The input distribution must have an independent copula.  
 
 See also
 --------
 WeightedExperiment
+TensorProductExperiment
```
Comment (josephmure) [inline python/src/GaussProductExperiment_doc.i.in:38]:
```suggestion
WeightedExperiment, TensorProductExperiment
```
This is the way the "See also" section is formatted in other classes.

File: python/src/TensorProductExperiment_doc.i.in [hunk 14]
```diff
@@ -119,6 +119,7 @@ and each node is equal to the agregation of the marginal elementary nodes:
 See also
 --------
 WeightedExperiment
+GaussProductExperiment
```
Comment (josephmure) [inline python/src/TensorProductExperiment_doc.i.in:122]:
```suggestion
WeightedExperiment, GaussProductExperiment
```
Same.

**Review by josephmure — CHANGES_REQUESTED** [review]

**Review by josephmure — APPROVED** [review]
Great!
